src/rgw/rgw_notify.cc

```diff
@@ -2,20 +2,576 @@
+        std::chrono::milliseconds(duration_jitter(rnd_gen));
+      timer.expires_from_now(duration);
+      const auto tp = ceph::coarse_real_time::clock::to_time_t(ceph::coarse_real_time::clock::now() + duration);
+      ldout(cct, 20) << "INFO: next queues processing will happen at: " << std::ctime(&tp)  << dendl;
+      boost::system::error_code ec;
+      timer.async_wait(yield[ec]);
+
+      queues_t queues;
+      auto ret = read_queue_list(queues, optional_yield(io_context, yield));
+      if (ret < 0) {
+        has_error = true;
+        continue;
+      }
+
+      std::vector<std::string> queue_gc;
+      std::mutex queue_gc_lock;
+      for (const auto& queue_name : queues) {
+        // try to lock the queue to check if it is owned by this rgw
+        // or if ownershif needs to be taken
+        librados::ObjectWriteOperation op;
+        op.assert_exists();
+        rados::cls::lock::lock(&op, queue_name+"_lock", 
+              ClsLockType::EXCLUSIVE,
+              lock_cookie, 
+              "" /*no tag*/,
+              "" /*no description*/,
+              failover_time,
+              LOCK_FLAG_MAY_RENEW);
+
+        ret = rgw_rados_operate(rados_ioctx, queue_name, &op, optional_yield(io_context, yield));
+        if (ret == -EBUSY) {
+          // lock is already taken by another RGW
+          ldout(cct, 20) << "INFO: queue: " << queue_name << " owned (locked) by another daemon" << dendl;
+          // if queue is owned, processing should be stopped
+          auto it = owned_queues.find(queue_name);
+          if (it != owned_queues.end() && it->second) {
+            it->second = false;
+            ldout(cct, 5) << "WARNING: queue: " << queue_name << " ownership lost. processing will stop" << dendl;
+          }
+          continue;
+        }
+        if (ret == -ENOENT) {
+          // queue is deleted - processing will stop the next time we try to read from the queue
+          ldout(cct, 10) << "INFO: queue: " << queue_name << " should not be locked - already deleted" << dendl;
+          continue;
+        }
+        if (ret < 0) {
+          // failed to lock for another reason, continue to process other queues
+          ldout(cct, 1) << "ERROR: failed to lock queue: " << queue_name << ". error: " << ret << dendl;
+          has_error = true;
+          continue;
+        }
+        // add queue to list of owned queues
+        const auto insert_result = owned_queues.insert(std::make_pair(queue_name, true));
+        if (insert_result.second) {
+          ldout(cct, 10) << "INFO: queue: " << queue_name << " now owned (locked) by this daemon" << dendl;
+          // start processing this queue
+          const auto& owned = insert_result.first->second;
+          spawn::spawn(io_context, [this, &queue_gc, &queue_gc_lock, &owned, queue_name](spawn::yield_context yield) {
+            process_queue(queue_name, owned, yield);
```

Comment:
is it thread-safe to share this `owned` reference across strands?

maybe it would be simpler for process_queue() and cleanup_queue() to add assert_locked() calls to their ops, so they get errors immediately if the lock expires or a conflicting lock is granted

Reply:
you are right, its not thread-safe (though, it won't crash either, as it will just access some garbage memory value and treat it as "boolean"...).
will add the lock check instead (also makes the code more readable).

Reply:
@cbodley fixed